packages/toolkit/src/query/core/buildThunks.ts

```diff
@@ -236,41 +237,58 @@ export function buildThunks<
   type State = RootState<any, string, ReducerPath>
 
   const patchQueryData: PatchQueryDataThunk<EndpointDefinitions, State> =
-    (endpointName, args, patches) => (dispatch) => {
+    (endpointName, args, patches, updateProvided) => (dispatch, getState) => {
       const endpointDefinition = endpointDefinitions[endpointName]
+
+      const queryCacheKey = serializeQueryArgs({
+        queryArgs: args,
+        endpointDefinition,
+        endpointName,
+      })
+
       dispatch(
-        api.internalActions.queryResultPatched({
-          queryCacheKey: serializeQueryArgs({
-            queryArgs: args,
-            endpointDefinition,
-            endpointName,
-          }),
-          patches,
-        })
+        api.internalActions.queryResultPatched({ queryCacheKey, patches })
+      )
+
+      if (!updateProvided) {
+        return
+      }
+
+      const newValue = api.endpoints[endpointName].select(args)(getState())
+
+      const providedTags = calculateProvidedBy(
```

Comment:
Hmm, I see the merit of "we have only one code path" for moving this into `patchQueryData`. 
Still, it has a significant downside: it will always dispatch two actions, whereas the approach I had suggested dispatches only one for an update and two for `undo` (which should happen far less often).

Every of those dispatches *could* result in a rerender - and as it stands, we cannot mark them as `batchable` since we need to read the updated state immediately after dispatch.

In that light, could you please keep it with the "mostly 1 dispatch, sometimes 2" approach?

Reply:
That was a brainfart, of course we can mark them batchable - batching will not delay reducer execution, but only the resulting rerender.

In that case, could you please make sure that `queryResultPatched` and `updateProvidedBy` are marked with `prepareAutoBatched` ?